### PR TITLE
Added shield to French, Uikraine, Belarusian and Dutch roads

### PR DESCRIPTION
**Pull request by VladiMihaylenko**

### CODE REVIEW

File: indexer/ftypes_matcher.cpp [hunk 1]
```diff
@@ -712,20 +712,112 @@ class RussiaRoadShieldParser : public RoadShieldParser
...
     if (s[0] == 'E' || s[0] == strings::UniChar(1045)) // Latin and cyrillic.
-      return RoadShield(RoadShieldType::Euro_Motorway, rawText);
+      return RoadShield(RoadShieldType::Generic_Green, rawText);
 
-    return RoadShield(RoadShieldType::Russia_Highway, rawText);
+    return RoadShield(RoadShieldType::Generic_Blue, rawText);
   }
 };
 
+class SimpleUnicodeRoadShieldParser : public RoadShieldParser
```
Comment (rokuz) [inline indexer/ftypes_matcher.cpp]: Можно сразу сделать дне для символов, а для строк. Если надо будет искать, например, двухбуквенные знаки

**Comment by VladiMihaylenko** [comment]
Можно действительно так
вт, 7 февр. 2017 г. в 13:45, Roman Kuznetsov <notifications@github.com>:

File: indexer/ftypes_matcher.cpp [hunk 2]
```diff
@@ -693,8 +693,8 @@ class UKRoadShieldParser : public SimpleRoadShieldParser
 {
 public:
   UKRoadShieldParser(std::string const & baseRoadNumber)
-    : SimpleRoadShieldParser(baseRoadNumber, {{'M', RoadShieldType::UK_Motorway},
-                                              {'A', RoadShieldType::UK_Highway}})
+    : SimpleRoadShieldParser(
```
Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: Could you please explain what all this stuff does here, in ftypes_matcher?

Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: ftypes_matcher is devoted to BaseChecker and its subclasses only, moreover, quite specific subclasses of BaseChecker should not be here.

File: indexer/ftypes_matcher.cpp [hunk 3]
```diff
@@ -712,20 +712,112 @@ class RussiaRoadShieldParser : public RoadShieldParser
...
+class SimpleUnicodeRoadShieldParser : public RoadShieldParser
+{
+public:
+  struct Name
+  {
+    Name(char simpleName, strings::UniChar const & unicodeName)
+      : m_simpleName(simpleName), m_unicodeName(unicodeName)
+    {
+    }
+    char m_simpleName;
```
Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: Blank line between ctor and fields, please.

File: indexer/ftypes_matcher.cpp [hunk 4]
```diff
@@ -712,20 +712,112 @@ class RussiaRoadShieldParser : public RoadShieldParser
...
+
+    for (auto const & p : m_types)
+    {
+      Name const & name = p.first;
+      RoadShieldType const type = p.second;
+
+      if (rawText.find(name.m_simpleName) != std::string::npos)
+        return RoadShield(type, rawText);
+
+      strings::UniString s = strings::MakeUniString(rawText);
```
Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: Please, add a test for this method when rawText is empty.

Comment (VladiMihaylenko) [inline indexer/ftypes_matcher.cpp]: Do we really need a test for this case? There is check for an empty string above?

Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: Okay, now there's no need in the test.

File: indexer/ftypes_matcher.cpp (hunk 2 above)
Comment (VladiMihaylenko) [inline indexer/ftypes_matcher.cpp]: I think now it's too much code for storing it in ftypes_matcher. 
I'll do a separate file for shield parsing.

Comment (ygorshenin) [inline indexer/ftypes_matcher.cpp]: It would be nice, thanks!

**Comment by VladiMihaylenko** [comment]
DUDE JTTP

**Comment by VladiMihaylenko** [comment]
@rokuz @ygorshenin ptal

File: drape_frontend/apply_feature_functors.hpp [hunk 5]
```diff
@@ -21,6 +21,11 @@ class SymbolRuleProto;
 
 //#define CALC_FILTERED_POINTS
 
+namespace ftypes
+{
+class RoadShield;
+}  // namespace ftypes
```
Comment (ygorshenin) [inline drape_frontend/apply_feature_functors.hpp]: Note: generally for namespaces in forward-declaration we do not write trailing "// namespace ABC" comment. Feel free to remove it.

File: indexer/indexer.pro [hunk 6]
```diff
@@ -57,6 +57,7 @@ SOURCES += \
     point_to_int64.cpp \
     postcodes_matcher.cpp \  # it's in indexer due to editor wich is in indexer and depends on postcodes_marcher
     rank_table.cpp \
+    road_shields_parser.cpp \
```
Comment (ygorshenin) [inline indexer/indexer.pro]: Do not forget to update CMakeLists too.

File: indexer/road_shields_parser.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+#include "indexer/feature.hpp"
+
+#include "base/string_utils.hpp"
+
+#include <unordered_map>
+#include <utility>
+
+namespace ftypes
+{
+class RoadShieldParser
```
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Please, put all these helper classes into anonymous namespace. Otherwise, it will be possible to reference them from other translation units.

File: indexer/road_shields_parser.cpp [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+        result.push_back(std::move(shield));
+    }
+    return result;
+  }
+
+protected:
+  std::string const m_baseRoadNumber;
+};
+
+uint32_t constexpr kMaxRoadShieldTextSize = 8;
```
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Could you please be so kind to move this constant to the top of the anonymous namespace, before all classes declarations?

File: indexer/road_shields_parser.cpp [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+class USRoadShieldParser : public RoadShieldParser
+{
+public:
+  USRoadShieldParser(std::string const & baseRoadNumber)
+  : RoadShieldParser(baseRoadNumber)
+  {}
+
+  RoadShield ParseRoadShield(std::string const & rawText) override
+  {
+    static std::vector<std::string> kStatesCode = {
```
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Do you really need a vector here? Why not std::array or even old good plain C array of strings? I know, this is not your code, so feel free to skip this comment at all.

Comment (VladiMihaylenko) [inline indexer/road_shields_parser.cpp]: Looks like it will be better if I replace it with std::array and move it into a class scope to avoid guards which are produced by local static constants.

Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Or, you even may move it to the global scope of the translation unit. In any case, feel free to do this in an individual pull request.

File: indexer/indexer.pro (hunk 6 above)
Comment (VladiMihaylenko) [inline indexer/indexer.pro]: Done

File: indexer/road_shields_parser.cpp [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    return RoadShield(RoadShieldType::Generic_Blue, rawText);
+  }
+};
+
+class SimpleUnicodeRoadShieldParser : public RoadShieldParser
+{
+public:
+  struct Name
+  {
+    Name(std::string simpleName, strings::UniString const & unicodeName)
```
Comment (rokuz) [inline indexer/road_shields_parser.cpp]: std::string const & simpleName

Comment (rokuz) [inline indexer/road_shields_parser.cpp]: + Юникодную можешь прямо здесь формировать strings::MakeUniString(simpleName)

File: indexer/road_shields_parser.cpp [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+    for (auto const & p : m_types)
+    {
+      Name const & name = p.first;
+      RoadShieldType const type = p.second;
+
+      if (rawText.find(name.m_simpleName) != std::string::npos)
+        return RoadShield(type, rawText);
+
+      auto const length = strings::ToUtf8(name.m_unicodeName).size();
```
Comment (rokuz) [inline indexer/road_shields_parser.cpp]: а у m_unicodeName размер нельзя получить? Без конвертации в Utf8

**Comment by VladiMihaylenko** [comment]
@ygorshenin ptal

File: indexer/road_shields_parser.cpp (hunk 11 above)
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Только кол-во байт, не кол-во символов. А подсчет кол-ва символов по факту эквивалентен конвертации в Utf8.

File: indexer/road_shields_parser.cpp [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+#include <utility>
+
+namespace
+{
+using ftypes::RoadShield;
+using ftypes::RoadShieldType;
+
+uint32_t constexpr kMaxRoadShieldTextSize = 8;
+
+static std::array<std::string, 63> const kStatesCode {
```
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: No need in static keyword here.

File: indexer/road_shields_parser.cpp [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+std::vector<RoadShield> GetRoadShields(FeatureType const & f)
+{
+  std::string const roadNumber = f.GetRoadNumber();
+  if (roadNumber.empty())
+    return std::vector<RoadShield>();
+
+  // Find out country name.
+  std::string mwmName = f.GetID().GetMwmName();
+  ASSERT_NOT_EQUAL(mwmName, FeatureID::kInvalidFileName, ());
+  auto const underlinePos = mwmName.find('_');
```
Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Can we use StartsWith() instead of this substr?

Comment (ygorshenin) [inline indexer/road_shields_parser.cpp]: Consider to write a helper method, say, ParseMwmName(), and use it here.

File: indexer/road_shields_parser.cpp [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+      }
+      if (!modifierFound)
+        return RoadShield();
+      }
+
+    if (shieldParts.size() <= 1)
+    {
+      return RoadShield(RoadShieldType::Default, rawText);
+    }
+    else
```
Comment (syershov) [inline indexer/road_shields_parser.cpp]: No need else after return.

File: indexer/road_shields_parser.cpp [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+        // Process cases like "US Loop 16".
+        if (!strings::is_number(shieldParts[1]) && strings::is_number(shieldParts[2]))
+        {
+          roadNumber = shieldParts[2];
+          additionalInfo = shieldParts[1];
+        }
+      }
+      if (roadType == "I")
+        return RoadShield(RoadShieldType::US_Interstate, roadNumber, additionalInfo);
+      else if (std::find(kStatesCode.begin(), kStatesCode.end(), shieldParts[0]) !=kStatesCode.end())
```
Comment (syershov) [inline indexer/road_shields_parser.cpp]: Here too.

File: indexer/road_shields_parser.cpp [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+  : RoadShieldParser(baseRoadNumber)
+  , m_types(types)
+  {}
+
+  RoadShield ParseRoadShield(std::string const & rawText) override
+  {
+    if (rawText.size() > kMaxRoadShieldBytesSize)
+      return RoadShield();
+
+      for (auto const & p : m_types)
```
Comment (syershov) [inline indexer/road_shields_parser.cpp]: Fix ident please.

**Comment by syershov** [comment]
Please take a look at Format checker.

**Comment by rokuz** [comment]
Please no any clanf-format to apply_feature_functors.cpp, any part of drape actually
